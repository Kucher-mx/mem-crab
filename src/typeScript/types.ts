type valueTypes = {M: number; N: number; X: number};
type stateTypes = {
    M: number;
    N: number;
    X: number;
    table: cellTypes[][];
    rowSum: rowSumTypes[];
    colAverage: number[];
};
type cellTypes = {id: string; amount: number; isHighlited: boolean};
type actionTypes = {table: cellTypes[][], consts: valueTypes}
type rowSumTypes = {rowSum: number, rowPersent: number, isHoveredSum: boolean, id: string}

export type {actionTypes,cellTypes,rowSumTypes,stateTypes, valueTypes};
