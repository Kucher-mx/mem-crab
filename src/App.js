import './App.css';

function App() {
  return (
    <div className="App">
      <div>memkrab</div>
    </div>
  );
}

export default App;
