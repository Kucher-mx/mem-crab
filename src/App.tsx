import * as React from "react";
import {Route, Routes} from "react-router-dom";

import MainPage from "./pages/main/Main.component";
import StartPage from "./pages/start/Start.component";
import withStyles from "isomorphic-style-loader/withStyles";
//@ts-ignore
import s from "./App.module.css";

function App() {
    return (
        <div className={s.App}>
            <Routes>
                <Route path="/" element={<StartPage />} />
                <Route path="/app/:m/:n/:x" element={<MainPage />} />
            </Routes>
        </div>
    );
}

export default withStyles(s)(App);
