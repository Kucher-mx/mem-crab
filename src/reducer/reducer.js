const initialState = {
    M: 0,
    N: 0,
    X: 0,
    table: [[], []],
}

const reducer = (state = initialState, actions) => {
    switch(actions.type){
        case 'SET_CONSTS':
            console.log(actions);
            return {
                M: actions.value.M,
                N: actions.value.N,
                X: actions.value.X,
            }
        case 'ADD_ROW':
            return { M: state.M + 1 }
        case 'REMOVE_ROW':
            return { M: state.M - 1 }
        case 'ADD_AMOUNT':
            const tableClone = state.table
            for(let arr of tableClone){
                for(let cell of arr){
                    if(cell.id === actions.idToChange){
                        cell.amount++
                    }
                }
            }
            return { table: tableClone }
        default:
            break;
    }

    return state;
}

export default reducer;