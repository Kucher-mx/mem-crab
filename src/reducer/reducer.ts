import types from "../actions/actionTypes";
import {actionTypes, cellTypes, rowSumTypes} from "../typeScript/types";
import {addRow, calcColAverage, calcRowSum} from "../utils/utils";

const initialState: {M: number; N: number; X: number; table: []; rowSum: rowSumTypes[]; colAverage: []} = {
    M: 0,
    N: 0,
    X: 0,
    table: [],
    rowSum: [],
    colAverage: [],
};

const reducer = (
    state: {
        M: number;
        N: number;
        X: number;
        table: cellTypes[][];
        rowSum: rowSumTypes[];
        colAverage: number[];
    } = initialState,
    actions: {type: string; value?: actionTypes; id?: string | string[]}
) => {
    switch (actions.type) {
        case types.ADD_AMOUNT:
            if (actions.value) {
                return {
                    ...state,
                    table: state.table.map((arr: cellTypes[]) => {
                        arr.map((cell: cellTypes) => {
                            if (cell.id === actions.id) {
                                cell.amount++;
                            }
                            return cell;
                        });
                        return arr;
                    }),
                    rowSum: calcRowSum(actions.value.consts.M, actions.value.consts.N, actions.value.table),
                    colAverage: calcColAverage(actions.value.consts.M, actions.value.consts.N, actions.value.table),
                };
            }
            return state;

        case types.HIGHLITE:
            return {
                ...state,
                table: state.table.map(arr => {
                    arr.map(el => {
                        if(actions.id?.includes(el.id)){
                            el.isHighlited = true;
                        }
                        return el;
                    });
                    return arr;
                }),
            };

        case types.UNHIGHLITE:
            return {
                ...state,
                table: state.table.map((arr: cellTypes[]) => {
                    arr.map((el: cellTypes) => {
                        if (el.isHighlited) {
                            el.isHighlited = false;
                        }
                        return el;
                    });
                    return arr;
                }),
            };

        case types.SET_CONSTS:
            if (actions.value) {
                return {
                    ...state,
                    table: actions.value.table,
                    rowSum: calcRowSum(actions.value.consts.M, actions.value.consts.N, actions.value.table),
                    colAverage: calcColAverage(actions.value.consts.M, actions.value.consts.N, actions.value.table),
                    M: actions.value.consts.M,
                    N: actions.value.consts.N,
                    X: actions.value.consts.X,
                };
            }
            return state;

        case types.ADD_ROW:
            const addRowTable = addRow(state.table, state.N);

            return {
                ...state,
                table: addRowTable,
                M: Number(state.M) + 1,
                rowSum: calcRowSum(state.M, state.N, addRowTable),
                colAverage: calcColAverage(state.M, state.N, addRowTable),
            };

        case types.REMOVE_ROW:
            if (state.table.length === 1) {
                return {...state};
            }

            const removeTable = state.table.filter((_, idx) => idx < state.table.length - 1);

            return {
                ...state,
                M: state.M - 1,
                table: removeTable,
                rowSum: calcRowSum(state.M, state.N, removeTable),
                colAverage: calcColAverage(state.M - 1, state.N, removeTable),
            };

        case types.HIGHlITE_SUM:
            return {
                ...state,
                rowSum: state.rowSum.map(item => {
                    if (item.id === actions.id) {
                        item.isHoveredSum = true;
                    }
                    return item;
                }),
            };

        case types.UNHIGHlITE_SUM:
            return {
                ...state,
                rowSum: state.rowSum.map(item => {
                    item.isHoveredSum = false;
                    return item;
                }),
            };

        default:
            return state;
    }
};

export default reducer;
