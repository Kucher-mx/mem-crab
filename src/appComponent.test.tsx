import React from "react";
import ReactDOM from "react-dom";
import ReactTestUtils from "react-dom/test-utils";
import {act as actUtils} from "react-dom/test-utils";
import {Provider} from "react-redux";
import {createStore} from "redux";
import reducer from "./reducer/reducer";
import {create, act} from "react-test-renderer";
import {stateTypes} from "./typeScript/types";
import StyleContext from "isomorphic-style-loader/StyleContext";
import App from "./App";
import {MemoryRouter} from "react-router-dom";

let container, store, component, createComponent, renderComponent;
const insertCss = (...styles) => {
    return {};
};

beforeAll(() => {
    jest.spyOn(console, "error").mockImplementation(jest.fn());
});

describe("testing app component: ", () => {
    let preloadedState: stateTypes;
    beforeAll(() => {
        global.fetch = jest.fn().mockImplementation(
            () =>
                new Promise(resolve => {
                    resolve({
                        json: () => ({
                            table: [
                                {
                                    id: "row_id_1",
                                    row: [
                                        {id: "el_id_1", rowId: "row_id_1"},
                                        {id: "el_id_2", rowId: "row_id_1"},
                                        {id: "el_id_3", rowId: "row_id_1"},
                                    ],
                                },
                                {
                                    id: "row_id_2",
                                    row: [
                                        {id: "el_id_4", rowId: "row_id_2"},
                                        {id: "el_id_5", rowId: "row_id_2"},
                                        {id: "el_id_6", rowId: "row_id_2"},
                                    ],
                                },
                            ],
                            cellsToHighlight: {},
                            amountObj: {
                                el_id_1: 1,
                                el_id_2: 2,
                                el_id_3: 3,
                                el_id_4: 4,
                                el_id_5: 5,
                                el_id_6: 6,
                            },
                            rowSum: [
                                {
                                    rowSum: 6,
                                    rowPercent: 28.6,
                                    isHoveredSum: false,
                                    id: "rowSum_id_1",
                                },
                                {
                                    rowSum: 15,
                                    rowPercent: 71.4,
                                    isHoveredSum: false,
                                    id: "rowSum_id_2",
                                },
                            ],
                            colAverage: [
                                {amount: 3, id: "colSum_id_1"},
                                {amount: 4, id: "colSum_id_2"},
                                {amount: 5, id: "colSum_id_3"},
                            ],
                            M: 2,
                            N: 3,
                            X: 2,
                        }),
                    });
                })
        );
    });
    afterAll(() => {
        //@ts-ignore
        global.fetch.mockClear();
        delete global.fetch;
    });

    beforeEach(() => {
        preloadedState = {
            table: [],
            rowSum: [],
            colAverage: [],
            M: null,
            N: null,
            X: null,
            amountObj: {},
            cellsToHighlight: {},
        };

        store = createStore(reducer, preloadedState);

        act(() => {
            component = create(
                <StyleContext.Provider value={{insertCss}}>
                    <Provider store={store}>
                        <MemoryRouter initialEntries={["/"]}>
                            <React.StrictMode>
                                <App />
                            </React.StrictMode>
                        </MemoryRouter>
                    </Provider>
                </StyleContext.Provider>
            );
        });
    });

    test("app tests preloaded state setup", () => {
        expect(store.getState().M).toEqual(null);
        expect(store.getState().N).toEqual(null);
        expect(store.getState().X).toEqual(null);

        expect(component.toJSON()).toMatchSnapshot();

        preloadedState = {
            table: [
                {
                    id: "row_id_1",
                    row: [
                        {id: "el_id_1", rowId: "row_id_1"},
                        {id: "el_id_2", rowId: "row_id_1"},
                        {id: "el_id_3", rowId: "row_id_1"},
                    ],
                },
                {
                    id: "row_id_2",
                    row: [
                        {id: "el_id_4", rowId: "row_id_2"},
                        {id: "el_id_5", rowId: "row_id_2"},
                        {id: "el_id_6", rowId: "row_id_2"},
                    ],
                },
            ],
            cellsToHighlight: {},
            amountObj: {
                el_id_1: 1,
                el_id_2: 2,
                el_id_3: 3,
                el_id_4: 4,
                el_id_5: 5,
                el_id_6: 6,
            },
            rowSum: [
                {
                    rowSum: 6,
                    rowPercent: 28.6,
                    isHoveredSum: false,
                    id: "rowSum_id_1",
                },
                {
                    rowSum: 15,
                    rowPercent: 71.4,
                    isHoveredSum: false,
                    id: "rowSum_id_2",
                },
            ],
            colAverage: [
                {amount: 3, id: "colSum_id_1"},
                {amount: 4, id: "colSum_id_2"},
                {amount: 5, id: "colSum_id_3"},
            ],
            M: 2,
            N: 3,
            X: 2,
        };

        store = createStore(reducer, preloadedState);
        actUtils(() => {
            component.update(
                <StyleContext.Provider value={{insertCss}}>
                    <Provider store={store}>
                        <MemoryRouter initialEntries={["/"]}>
                            <React.StrictMode>
                                <App />
                            </React.StrictMode>
                        </MemoryRouter>
                    </Provider>
                </StyleContext.Provider>
            );
        });
        expect(preloadedState).toEqual(store.getState());

        expect(component.toJSON()).toMatchSnapshot();
    });

    test("send params", async () => {
        const instance = component.root;

        const M = instance.findByProps({id: "M"});
        const N = instance.findByProps({id: "N"});
        const X = instance.findByProps({id: "X"});

        expect(M.props.value).toEqual("");

        const mEvent = {currentTarget: {value: "2", name: "M"}};
        const nEvent = {currentTarget: {value: "3", name: "N"}};
        const xEvent = {currentTarget: {value: "2", name: "X"}};

        actUtils(() => {
            M.props.onChange(mEvent);
        });
        actUtils(() => {
            N.props.onChange(nEvent);
        });
        actUtils(() => {
            X.props.onChange(xEvent);
        });
        const form = instance.findByProps({id: "startForm"});
        const formEvent = {preventDefault: () => {}};

        await actUtils(async () => await form.props.onSubmit(formEvent));

        expect(store.getState().M).toEqual(2);
        expect(store.getState().N).toEqual(3);
        expect(store.getState().X).toEqual(2);

        expect(store.getState().table.length).toBe(2);
        expect(store.getState().table[0].row.length).toBe(3);
    });

    test("validate params in the form", () => {
        const instance = component.root;

        const M = instance.findByProps({id: "M"});
        const N = instance.findByProps({id: "N"});
        const X = instance.findByProps({id: "X"});

        const mEvent = {currentTarget: {value: "-1", name: "M"}};
        const nEvent = {currentTarget: {value: "3", name: "N"}};
        const xEvent = {currentTarget: {value: "2", name: "X"}};

        act(() => {
            M.props.onChange(mEvent);
        });
        act(() => {
            N.props.onChange(nEvent);
        });
        act(() => {
            X.props.onChange(xEvent);
        });

        const form = instance.findByProps({id: "startForm"});
        const formEvent = {preventDefault: () => {}};

        act(() => form.props.onSubmit(formEvent));

        expect(M.props.value).toBe("");
        expect(X.props.value).toBe("");
        expect(N.props.value).toBe("");
    });
});

describe("Testing mouse events: ", () => {
    const preloadedState = {
        table: [
            {
                id: "row_id_1",
                row: [
                    {id: "el_id_1", rowId: "row_id_1"},
                    {id: "el_id_2", rowId: "row_id_1"},
                    {id: "el_id_3", rowId: "row_id_1"},
                ],
            },
            {
                id: "row_id_2",
                row: [
                    {id: "el_id_4", rowId: "row_id_2"},
                    {id: "el_id_5", rowId: "row_id_2"},
                    {id: "el_id_6", rowId: "row_id_2"},
                ],
            },
        ],
        cellsToHighlight: {},
        amountObj: {
            el_id_1: 1,
            el_id_2: 2,
            el_id_3: 3,
            el_id_4: 4,
            el_id_5: 5,
            el_id_6: 6,
        },
        rowSum: [
            {
                rowSum: 6,
                rowPercent: 28.6,
                isHoveredSum: false,
                id: "rowSum_id_1",
            },
            {
                rowSum: 15,
                rowPercent: 71.4,
                isHoveredSum: false,
                id: "rowSum_id_2",
            },
        ],
        colAverage: [
            {amount: 3, id: "colSum_id_1"},
            {amount: 4, id: "colSum_id_2"},
            {amount: 5, id: "colSum_id_3"},
        ],
        M: 2,
        N: 3,
        X: 2,
    };

    container = document.createElement("div");
    document.body.appendChild(container);

    beforeEach(() => {
        store = createStore(reducer, preloadedState);

        act(() => {
            component = create(
                <StyleContext.Provider value={{insertCss}}>
                    <Provider store={store}>
                        <MemoryRouter initialEntries={["/app/2/3/2"]}>
                            <React.StrictMode>
                                <App />
                            </React.StrictMode>
                        </MemoryRouter>
                    </Provider>
                </StyleContext.Provider>
            );
        });

        actUtils(() => {
            renderComponent = ReactDOM.render(
                <StyleContext.Provider value={{insertCss}}>
                    <Provider store={store}>
                        <MemoryRouter initialEntries={["/app/2/3/2"]}>
                            <React.StrictMode>
                                <App />
                            </React.StrictMode>
                        </MemoryRouter>
                    </Provider>
                </StyleContext.Provider>,
                container
            );
        });
    });

    test("show table and handle ceil`s increase", () => {
        const instance = component.root;

        const ceil = instance.findByProps({id: "el_id_2"});

        const ceilDom = container.querySelector("#el_id_2");

        actUtils(() => ReactTestUtils.Simulate.click(ceilDom));
        const amount = store.getState().amountObj["el_id_2"];

        expect(+ceil.props.amount).toEqual(amount);
        // expect(component.toJSON()).toMatchSnapshot();
    });

    test("test hover on cell", () => {
        const ceil = container.querySelector("#el_id_2");

        actUtils(() => ReactTestUtils.Simulate.mouseEnter(ceil));
        const afterIn = store.getState().cellsToHighlight.row_id_1.el_id_2;
        expect(afterIn).toBe(true);

        actUtils(() => ReactTestUtils.Simulate.mouseLeave(ceil));
        const afterOut = store.getState().cellsToHighlight.row_id_1;
        expect(afterOut).toBeFalsy();

        expect(component.toJSON()).toMatchSnapshot();
    });

    test("test hover on sum element", () => {
        const sumCeil = container.querySelector("#rowSum_id_1");

        actUtils(() => ReactTestUtils.Simulate.mouseEnter(sumCeil));
        const afterEnter = store.getState().rowSum[0].isHoveredSum;
        expect(afterEnter).toBe(true);

        actUtils(() => ReactTestUtils.Simulate.mouseLeave(sumCeil));
        const afterLeave = store.getState().rowSum[0].isHoveredSum;
        expect(afterLeave).toBe(false);

        expect(component.toJSON()).toMatchSnapshot();
    });

    test("test adding row to the end button", () => {
        const rowsAmount = store.getState().table.length;
        const addButton = container.querySelector("#addRow");
        expect(component.toJSON()).toMatchSnapshot();

        actUtils(() => ReactTestUtils.Simulate.click(addButton));
        const lengthAfterAdd = store.getState().table.length;

        expect(rowsAmount + 1).toBe(lengthAfterAdd);
    });

    test("test removing row button", () => {
        const rowsAmount = store.getState().table.length;
        const removeButton = container.querySelector("#removeRow");

        actUtils(() => ReactTestUtils.Simulate.click(removeButton));
        const lengthAfterRemove = store.getState().table.length;

        expect(rowsAmount - 1).toBe(lengthAfterRemove);
        // expect(component.toJSON()).toMatchSnapshot();
    });

    test("test removing last row rows", () => {
        const rowsAmount = store.getState().table.length;
        const removeButton = container.querySelector("#removeRow");

        actUtils(() => ReactTestUtils.Simulate.click(removeButton));
        actUtils(() => ReactTestUtils.Simulate.click(removeButton));
        const lengthAfterRemove = store.getState().table.length;

        expect(rowsAmount - 1).toBe(lengthAfterRemove);
        // expect(component.toJSON()).toMatchSnapshot();
    });
});
