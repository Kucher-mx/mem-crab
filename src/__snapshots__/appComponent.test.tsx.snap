// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing mouse events:  test adding row to the end button 1`] = `
<div>
  <div>
    <div>
      <a
        href="/"
        onClick={[Function]}
      >
        MEM-CRAB React Test
      </a>
      <div>
        <table>
          <tbody>
            <tr>
              <td
                className=""
                id="el_id_1"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
              >
                1
              </td>
              <td
                className=""
                id="el_id_2"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
              >
                2
              </td>
              <td
                className=""
                id="el_id_3"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
              >
                3
              </td>
            </tr>
            <tr>
              <td
                className=""
                id="el_id_4"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
              >
                4
              </td>
              <td
                className=""
                id="el_id_5"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
              >
                5
              </td>
              <td
                className=""
                id="el_id_6"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
              >
                6
              </td>
            </tr>
          </tbody>
        </table>
        <div>
          <div
            id="rowSum_id_1"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <div
              style={
                Object {
                  "width": "0%",
                }
              }
            />
            6
          </div>
          <div
            id="rowSum_id_2"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <div
              style={
                Object {
                  "width": "0%",
                }
              }
            />
            15
          </div>
        </div>
      </div>
      <div>
        <div
          style={
            Object {
              "gridTemplateColumns": "repeat(3, 1fr)",
            }
          }
        >
          <div>
            3
          </div>
          <div>
            4
          </div>
          <div>
            5
          </div>
        </div>
        <div>
          21
        </div>
      </div>
      <div>
        <button
          id="addRow"
          onClick={[Function]}
          type="button"
        >
          +
        </button>
        <button
          id="removeRow"
          onClick={[Function]}
          type="button"
        >
          -
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Testing mouse events:  test hover on cell 1`] = `
<div>
  <div>
    <div>
      <a
        href="/"
        onClick={[Function]}
      >
        MEM-CRAB React Test
      </a>
      <div>
        <table>
          <tbody>
            <tr>
              <td
                className=""
                id="el_id_1"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
              >
                1
              </td>
              <td
                className=""
                id="el_id_2"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
              >
                2
              </td>
              <td
                className=""
                id="el_id_3"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
              >
                3
              </td>
            </tr>
            <tr>
              <td
                className=""
                id="el_id_4"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
              >
                4
              </td>
              <td
                className=""
                id="el_id_5"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
              >
                5
              </td>
              <td
                className=""
                id="el_id_6"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
              >
                6
              </td>
            </tr>
          </tbody>
        </table>
        <div>
          <div
            id="rowSum_id_1"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <div
              style={
                Object {
                  "width": "0%",
                }
              }
            />
            6
          </div>
          <div
            id="rowSum_id_2"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <div
              style={
                Object {
                  "width": "0%",
                }
              }
            />
            15
          </div>
        </div>
      </div>
      <div>
        <div
          style={
            Object {
              "gridTemplateColumns": "repeat(3, 1fr)",
            }
          }
        >
          <div>
            3
          </div>
          <div>
            4
          </div>
          <div>
            5
          </div>
        </div>
        <div>
          21
        </div>
      </div>
      <div>
        <button
          id="addRow"
          onClick={[Function]}
          type="button"
        >
          +
        </button>
        <button
          id="removeRow"
          onClick={[Function]}
          type="button"
        >
          -
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Testing mouse events:  test hover on sum element 1`] = `
<div>
  <div>
    <div>
      <a
        href="/"
        onClick={[Function]}
      >
        MEM-CRAB React Test
      </a>
      <div>
        <table>
          <tbody>
            <tr>
              <td
                className=""
                id="el_id_1"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
              >
                1
              </td>
              <td
                className=""
                id="el_id_2"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
              >
                2
              </td>
              <td
                className=""
                id="el_id_3"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
              >
                3
              </td>
            </tr>
            <tr>
              <td
                className=""
                id="el_id_4"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
              >
                4
              </td>
              <td
                className=""
                id="el_id_5"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
              >
                5
              </td>
              <td
                className=""
                id="el_id_6"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
              >
                6
              </td>
            </tr>
          </tbody>
        </table>
        <div>
          <div
            id="rowSum_id_1"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <div
              style={
                Object {
                  "width": "0%",
                }
              }
            />
            6
          </div>
          <div
            id="rowSum_id_2"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <div
              style={
                Object {
                  "width": "0%",
                }
              }
            />
            15
          </div>
        </div>
      </div>
      <div>
        <div
          style={
            Object {
              "gridTemplateColumns": "repeat(3, 1fr)",
            }
          }
        >
          <div>
            3
          </div>
          <div>
            4
          </div>
          <div>
            5
          </div>
        </div>
        <div>
          21
        </div>
      </div>
      <div>
        <button
          id="addRow"
          onClick={[Function]}
          type="button"
        >
          +
        </button>
        <button
          id="removeRow"
          onClick={[Function]}
          type="button"
        >
          -
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`testing app component:  app tests preloaded state setup 1`] = `
<div>
  <div>
    <form
      id="startForm"
      onSubmit={[Function]}
    >
      <div>
        <label>
          M:
          <input
            defaultValue={1}
            id="M"
            name="M"
            onChange={[Function]}
            required={true}
            type="number"
            value=""
          />
        </label>
        <label>
          N:
          <input
            defaultValue={3}
            id="N"
            name="N"
            onChange={[Function]}
            required={true}
            type="number"
            value=""
          />
        </label>
        <label>
          X:
          <input
            defaultValue={1}
            id="X"
            name="X"
            onChange={[Function]}
            required={true}
            type="number"
            value=""
          />
        </label>
      </div>
      <button
        type="submit"
      >
        submit
      </button>
    </form>
  </div>
</div>
`;

exports[`testing app component:  app tests preloaded state setup 2`] = `
<div>
  <div>
    <form
      id="startForm"
      onSubmit={[Function]}
    >
      <div>
        <label>
          M:
          <input
            defaultValue={1}
            id="M"
            name="M"
            onChange={[Function]}
            required={true}
            type="number"
            value=""
          />
        </label>
        <label>
          N:
          <input
            defaultValue={3}
            id="N"
            name="N"
            onChange={[Function]}
            required={true}
            type="number"
            value=""
          />
        </label>
        <label>
          X:
          <input
            defaultValue={1}
            id="X"
            name="X"
            onChange={[Function]}
            required={true}
            type="number"
            value=""
          />
        </label>
      </div>
      <button
        type="submit"
      >
        submit
      </button>
    </form>
  </div>
</div>
`;
